File: backend/services/interview_turns.py
"""
面试原文 → 结构化 turns

每个 turn 是一次发言（speaker 切换处切分）。给每个 turn 全局唯一 id，
后续 LLM 引用 id 而不是再复制原文片段，前端按 id 定位。

输出结构：
    [
        {"id": 0, "speaker": "面试官", "content": "...", "char_start": 0, "char_end": 12},
        {"id": 1, "speaker": "我", "content": "...", "char_start": 13, "char_end": 80},
        ...
    ]

切分规则：
    - 以行首 "面试官: / 面试官： / 我: / 我：" 作为 turn 边界（前后可有空白）
    - 该行的说话人前缀本身不进入 content
    - 找不到任何说话人标记时，按双换行切段，speaker 留空
"""
from __future__ import annotations

import re

# 行首说话人前缀：面试官 / 我（兼容中英文冒号、前后空白）
_SPEAKER_RE = re.compile(r"(?:^|\n)([ \t]*)(面试官|我)\s*[:：]\s*", re.MULTILINE)


def split_into_turns(text: str) -> list[dict]:
    """把原文切成 turns 列表（带全局 id 和原文字符偏移）。"""
    if not text:
        return []

    matches = list(_SPEAKER_RE.finditer(text))
    if not matches:
        # 没有说话人标记 → 按空行切段，speaker 留空，便于纯笔记式输入也能定位
        return _split_by_blank_lines(text)

    turns: list[dict] = []
    for i, m in enumerate(matches):
        # content 起点：匹配末尾（说话人前缀之后）
        content_start = m.end()
        # content 终点：下一个匹配的起点（或文末）
        content_end = matches[i + 1].start() if i + 1 < len(matches) else len(text)
        content = text[content_start:content_end].rstrip()
        if not content:
            continue
        turns.append({
            "id": len(turns),
            "speaker": m.group(2),
            "content": content,
            "char_start": content_start,
            "char_end": content_start + len(content),
        })
    return turns or _split_by_blank_lines(text)


def _split_by_blank_lines(text: str) -> list[dict]:
    """fallback：按双换行切段，speaker 留空。"""
    turns: list[dict] = []
    cursor = 0
    for part in re.split(r"\n\s*\n", text):
        stripped = part.strip()
        if not stripped:
            cursor += len(part) + 2  # 近似估算，无需精确
            continue
        start = text.find(stripped, cursor)
        if start < 0:
            start = cursor
        end = start + len(stripped)
        turns.append({
            "id": len(turns),
            "speaker": "",
            "content": stripped,
            "char_start": start,
            "char_end": end,
        })
        cursor = end
    if not turns:
        turns.append({
            "id": 0,
            "speaker": "",
            "content": text.strip(),
            "char_start": 0,
            "char_end": len(text),
        })
    return turns


# ============================================================
# ASR 切错修复（启发式合并）
# ============================================================

# 续接字：基本只能跟在另一字后面，单独开句无意义。命中即与上一 turn 合并。
_CONT_START = ("种", "吗", "呢", "啊", "呀", "哈", "嘛", "咯")
# 上一 turn 结尾的"悬挂连词"：命中说明被腰斩
_HANG_END = ("那", "和", "与", "或", "及", "就", "且")
# 句末标点（含中英文）
_TERMINATORS = set("。？！.?!；;")
# 修复时允许的"短碎片"长度
_SHORT_FRAG_LEN = 8


def _should_merge_to_prev(prev: dict | None, cur: dict) -> bool:
    """判定当前 turn 是否应当合并到上一个 turn（speaker 沿用上一个）。

    保守策略：只在以下任一强信号命中时合并：
    1) 当前 turn 以续接字开头（如「种」「吗」「呢」），且当前 turn 较短（< 8 字）。
    2) 上一个 turn 以悬挂连词结尾（如「那」「和」「或」），且当前 turn 较短无句末标点。
    """
    if not prev:
        return False
    cur_content = (cur.get("content") or "").strip()
    if not cur_content:
        return False

    prev_content = (prev.get("content") or "").rstrip("，,。.？?！!；;、 \t\n")
    if not prev_content:
        return False
    prev_last = prev_content[-1]
    cur_first = cur_content[0]

    # 信号 1：续接字开头 + 当前短
    if cur_first in _CONT_START and len(cur_content) < _SHORT_FRAG_LEN:
        return True

    # 信号 2：上句悬挂连词结尾 + 当前短 + 无句末标点
    if (
        prev_last in _HANG_END
        and len(cur_content) < _SHORT_FRAG_LEN
        and not any(c in _TERMINATORS for c in cur_content)
    ):
        return True

    return False


def repair_turns(turns: list[dict]) -> list[dict]:
    """合并被 ASR 切错的破碎 turn 到上一个，speaker 沿用上一个。

    - 重新连续编号 id，char_start/char_end 取合并后的首末。
    - 输入空返回空。
    - 保守合并：见 _should_merge_to_prev。
    """
    if not turns:
        return turns
    out: list[dict] = []
    for t in turns:
        prev = out[-1] if out else None
        if _should_merge_to_prev(prev, t):
            prev["content"] = prev["content"].rstrip() + t["content"].lstrip()
            if "char_end" in t:
                prev["char_end"] = t["char_end"]
        else:
            out.append(dict(t))
    # 重新连续编号
    for i, t in enumerate(out):
        t["id"] = i
    return out


def render_turns_for_llm(turns: list[dict]) -> str:
    """把 turns 渲染回带 [tN] 标记的对话，供 LLM 解析。

    输出形如：
        [t0] 面试官: ...
        [t1] 我: ...

    保留 speaker 前缀让 LLM 仍能从文本特征判定说话人；turn id 让它能直接引用。
    """
    lines: list[str] = []
    for t in turns:
        speaker = t.get("speaker") or ""
        prefix = f"{speaker}: " if speaker else ""
        lines.append(f"[t{t['id']}] {prefix}{t['content']}")
    return "\n".join(lines)


def chunk_turns(turns: list[dict], chunk_size: int = 1200) -> list[list[dict]]:
    """把 turns 切成多段，每段累计字符不超过 chunk_size。
    单 turn 超长时它自成一段。
    """
    if not turns:
        return []
    chunks: list[list[dict]] = []
    current: list[dict] = []
    current_len = 0
    for t in turns:
        t_len = len(t["content"]) + 8  # 算上 [tN] speaker: 前缀的大致开销
        if not current and t_len > chunk_size:
            chunks.append([t])
            continue
        if current_len + t_len > chunk_size and current:
            chunks.append(current)
            current = [t]
            current_len = t_len
        else:
            current.append(t)
            current_len += t_len
    if current:
        chunks.append(current)
    return chunks

File: java-backend/src/main/java/com/interview/agent/interview/support/InterviewTurns.java
package com.interview.agent.interview.support;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 面试原文 → 结构化 turns —— 完全复刻 Python {@code backend/services/interview_turns.py}。
 *
 * <p>每个 turn 是一次发言（speaker 切换处切分），带全局唯一 id 与原文字符偏移。
 * 后续 LLM 引用 id 而不再复制原文片段。</p>
 *
 * <p>复刻要点（不得改动）：
 * <ul>
 *   <li>说话人正则：{@code (?:^|\n)([ \t]*)(面试官|我)\s*[:：]\s*}</li>
 *   <li>说话人前缀不进入 content；content 末尾 rstrip</li>
 *   <li>无说话人标记 → 按双换行切段，speaker 留空</li>
 *   <li>repair 续接字 / 悬挂连词常量逐字一致；短碎片阈值 8</li>
 *   <li>chunk t_len = content 长度 + 8（[tN] speaker: 前缀开销）</li>
 * </ul>
 */
public final class InterviewTurns {

    private InterviewTurns() {
    }

    public static final int DEFAULT_CHUNK_SIZE = 1200;

    /** 行首说话人前缀：面试官 / 我（兼容中英文冒号、前后空白）。 */
    private static final Pattern SPEAKER_RE = Pattern.compile("(?:^|\\n)([ \\t]*)(面试官|我)\\s*[:：]\\s*");
    /** 按双换行切段。 */
    private static final Pattern BLANK_LINE_RE = Pattern.compile("\\n\\s*\\n");

    /** 续接字：基本只能跟在另一字后面，单独开句无意义。命中即与上一 turn 合并。 */
    private static final Set<Character> CONT_START = Set.of('种', '吗', '呢', '啊', '呀', '哈', '嘛', '咯');
    /** 上一 turn 结尾的"悬挂连词"：命中说明被腰斩。 */
    private static final Set<Character> HANG_END = Set.of('那', '和', '与', '或', '及', '就', '且');
    /** 句末标点（含中英文）。 */
    private static final Set<Character> TERMINATORS = Set.of('。', '？', '！', '.', '?', '!', '；', ';');
    /** 修复时允许的"短碎片"长度。 */
    private static final int SHORT_FRAG_LEN = 8;
    /** prev 末尾裁剪字符集（与 Python rstrip 参数一致）。 */
    private static final String PREV_TRIM = "，,。.？?！!；;、 \t\n";

    // ============================================================
    // 切分
    // ============================================================

    /** 把原文切成 turns 列表（带全局 id 和原文字符偏移）。 */
    public static List<Map<String, Object>> splitIntoTurns(String text) {
        if (text == null || text.isEmpty()) {
            return new ArrayList<>();
        }
        Matcher m = SPEAKER_RE.matcher(text);
        List<int[]> spans = new ArrayList<>();   // [wholeStart, wholeEnd]
        List<String> speakers = new ArrayList<>();
        while (m.find()) {
            spans.add(new int[]{m.start(), m.end()});
            speakers.add(m.group(2));
        }
        if (spans.isEmpty()) {
            return splitByBlankLines(text);
        }

        List<Map<String, Object>> turns = new ArrayList<>();
        for (int i = 0; i < spans.size(); i++) {
            int contentStart = spans.get(i)[1];
            int contentEnd = (i + 1 < spans.size()) ? spans.get(i + 1)[0] : text.length();
            String content = rstripAll(text.substring(contentStart, contentEnd));
            if (content.isEmpty()) {
                continue;
            }
            Map<String, Object> turn = new LinkedHashMap<>();
            turn.put("id", turns.size());
            turn.put("speaker", speakers.get(i));
            turn.put("content", content);
            turn.put("char_start", contentStart);
            turn.put("char_end", contentStart + content.length());
            turns.add(turn);
        }
        return turns.isEmpty() ? splitByBlankLines(text) : turns;
    }

    /** fallback：按双换行切段，speaker 留空。 */
    private static List<Map<String, Object>> splitByBlankLines(String text) {
        List<Map<String, Object>> turns = new ArrayList<>();
        int cursor = 0;
        for (String part : BLANK_LINE_RE.split(text)) {
            String stripped = part.strip();
            if (stripped.isEmpty()) {
                cursor += part.length() + 2;
                continue;
            }
            int start = text.indexOf(stripped, cursor);
            if (start < 0) {
                start = cursor;
            }
            int end = start + stripped.length();
            Map<String, Object> turn = new LinkedHashMap<>();
            turn.put("id", turns.size());
            turn.put("speaker", "");
            turn.put("content", stripped);
            turn.put("char_start", start);
            turn.put("char_end", end);
            turns.add(turn);
            cursor = end;
        }
        if (turns.isEmpty()) {
            Map<String, Object> turn = new LinkedHashMap<>();
            turn.put("id", 0);
            turn.put("speaker", "");
            turn.put("content", text.strip());
            turn.put("char_start", 0);
            turn.put("char_end", text.length());
            turns.add(turn);
        }
        return turns;
    }

    // ============================================================
    // ASR 切错修复（启发式合并）
    // ============================================================

    /** 合并被 ASR 切错的破碎 turn 到上一个，speaker 沿用上一个；最后重新连续编号 id。 */
    public static List<Map<String, Object>> repairTurns(List<Map<String, Object>> turns) {
        if (turns == null || turns.isEmpty()) {
            return turns;
        }
        List<Map<String, Object>> out = new ArrayList<>();
        for (Map<String, Object> t : turns) {
            Map<String, Object> prev = out.isEmpty() ? null : out.get(out.size() - 1);
            if (shouldMergeToPrev(prev, t)) {
                String mergedContent = rstripAll(str(prev.get("content"))) + lstripAll(str(t.get("content")));
                prev.put("content", mergedContent);
                if (t.containsKey("char_end")) {
                    prev.put("char_end", t.get("char_end"));
                }
            } else {
                out.add(new LinkedHashMap<>(t));
            }
        }
        for (int i = 0; i < out.size(); i++) {
            out.get(i).put("id", i);
        }
        return out;
    }

    /** 判定当前 turn 是否应合并到上一个 turn（保守策略，与 Python 一致）。 */
    private static boolean shouldMergeToPrev(Map<String, Object> prev, Map<String, Object> cur) {
        if (prev == null) {
            return false;
        }
        String curContent = str(cur.get("content")).strip();
        if (curContent.isEmpty()) {
            return false;
        }
        String prevContent = stripTrailingChars(str(prev.get("content")), PREV_TRIM);
        if (prevContent.isEmpty()) {
            return false;
        }
        char prevLast = prevContent.charAt(prevContent.length() - 1);
        char curFirst = curContent.charAt(0);

        // 信号 1：续接字开头 + 当前短
        if (CONT_START.contains(curFirst) && curContent.length() < SHORT_FRAG_LEN) {
            return true;
        }
        // 信号 2：上句悬挂连词结尾 + 当前短 + 无句末标点
        if (HANG_END.contains(prevLast) && curContent.length() < SHORT_FRAG_LEN
                && !containsAny(curContent, TERMINATORS)) {
            return true;
        }
        return false;
    }

    // ============================================================
    // 渲染 / 分块
    // ============================================================

    /** 把 turns 渲染回带 [tN] 标记的对话，供 LLM 解析。 */
    public static String renderTurnsForLlm(List<Map<String, Object>> turns) {
        List<String> lines = new ArrayList<>();
        for (Map<String, Object> t : turns) {
            String speaker = str(t.get("speaker"));
            String prefix = speaker.isEmpty() ? "" : speaker + ": ";
            lines.add("[t" + t.get("id") + "] " + prefix + str(t.get("content")));
        }
        return String.join("\n", lines);
    }

    /** 把 turns 切成多段，每段累计字符不超过 chunkSize；单 turn 超长时自成一段。 */
    public static List<List<Map<String, Object>>> chunkTurns(List<Map<String, Object>> turns, int chunkSize) {
        List<List<Map<String, Object>>> chunks = new ArrayList<>();
        if (turns == null || turns.isEmpty()) {
            return chunks;
        }
        List<Map<String, Object>> current = new ArrayList<>();
        int currentLen = 0;
        for (Map<String, Object> t : turns) {
            int tLen = str(t.get("content")).length() + 8;  // [tN] speaker: 前缀大致开销
            if (current.isEmpty() && tLen > chunkSize) {
                chunks.add(new ArrayList<>(List.of(t)));
                continue;
            }
            if (currentLen + tLen > chunkSize && !current.isEmpty()) {
                chunks.add(current);
                current = new ArrayList<>();
                current.add(t);
                currentLen = tLen;
            } else {
                current.add(t);
                currentLen += tLen;
            }
        }
        if (!current.isEmpty()) {
            chunks.add(current);
        }
        return chunks;
    }

    // ============================================================
    // 小工具
    // ============================================================

    private static String str(Object o) {
        return o == null ? "" : o.toString();
    }

    /** 去掉尾部所有空白（等价 Python str.rstrip()）。 */
    private static String rstripAll(String s) {
        int end = s.length();
        while (end > 0 && Character.isWhitespace(s.charAt(end - 1))) {
            end--;
        }
        return s.substring(0, end);
    }

    /** 去掉头部所有空白（等价 Python str.lstrip()）。 */
    private static String lstripAll(String s) {
        int start = 0;
        while (start < s.length() && Character.isWhitespace(s.charAt(start))) {
            start++;
        }
        return s.substring(start);
    }

    /** 去掉尾部指定字符集（等价 Python str.rstrip(chars)）。 */
    private static String stripTrailingChars(String s, String chars) {
        int end = s.length();
        while (end > 0 && chars.indexOf(s.charAt(end - 1)) >= 0) {
            end--;
        }
        return s.substring(0, end);
    }

    private static boolean containsAny(String s, Set<Character> charset) {
        for (int i = 0; i < s.length(); i++) {
            if (charset.contains(s.charAt(i))) {
                return true;
            }
        }
        return false;
    }
}
